cpp/include/cuvs/neighbors/all_neighbors.hpp

```diff
@@ -0,0 +1,152 @@
+/*
+ * Copyright (c) 2025, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cuvs/neighbors/ivf_pq.hpp>
+#include <cuvs/neighbors/nn_descent.hpp>
+
+#include <variant>
+
+namespace cuvs::neighbors::all_neighbors {
+
+/**
+ * @brief Parameters used to build an all-neighbors knn graph (find nearest neighbors for all the
+ * training vectors)
+ */
+namespace graph_build_params {
+
+/** Specialized parameters utilizing IVF-PQ to build knn graph */
+struct ivf_pq_params {
+  cuvs::neighbors::ivf_pq::index_params build_params;
+  cuvs::neighbors::ivf_pq::search_params search_params;
+  float refinement_rate = 2.0;
+};
+
+using nn_descent_params = cuvs::neighbors::nn_descent::index_params;
+}  // namespace graph_build_params
+
+using GraphBuildParams =
+  std::variant<graph_build_params::ivf_pq_params, graph_build_params::nn_descent_params>;
+
+/**
+ * @brief Parameters used to build an all-neighbors graph  (find nearest neighbors for all the
+ * training vectors)
+ *
+ * graph_build_params: graph building parameters for the given graph building algorithm. defaults
+ * to ivfpq.
+ * n_nearest_clusters: number of nearest clusters each data point will be assigned to in
+ * the batching algorithm
+ * n_clusters: number of total clusters (aka batches) to split the data into. If set to 1, algorithm
+ * creates an all-neighbors graph without batching
+ * metric: metric type
+ *
+ */
+struct all_neighbors_params {
+  /** Parameters for knn graph building algorithm
+   *
+   * Set ivf_pq_params, or nn_descent_params to select the graph build
+   * algorithm and control their parameters.
+   *
+   * @code{.cpp}
+   * all_neighbors::index_params params;
+   * // 1. Choose IVF-PQ algorithm
+   * params.graph_build_params = all_neighbors::graph_build_params::ivf_pq_params{};
+   *
+   * // 2. Choose NN Descent algorithm for kNN graph construction
+   * params.graph_build_params = all_neighbors::graph_build_params::nn_descent_params{};
+   *
+   * @endcode
+   */
+  GraphBuildParams graph_build_params;
+
+  /**
+   * Usage of n_nearest_clusters and n_clusters
+   *
+   * Hint1: the ratio of n_nearest_clusters / n_clusters determines device memory usage.
```

Comment:
Instead of `hint1` and `hint2`, can we just put this in a paragraph form? the `hint` make it sound like a video game.

Reply:
hahaha 😂 sure, will do